simulation/tests/test_agents.py

```diff
@@ -150,3 +150,60 @@ def test_household_gets_non_zero_insulation_quotes_for_all_upgradable_elements(
 
         assert set(insulation_quotes.keys()) == upgradable_elements
         assert all(quote > 0 for quote in insulation_quotes.values())
+
+    def test_household_chooses_one_to_three_insulation_measures_to_install(
+        self,
+    ) -> None:
+
+        household = household_factory()
+        assert 0 < household.choose_n_elements_to_insulate() <= 3
+
+    def test_household_chooses_cheapest_insulation_measures(
+        self,
+    ) -> None:
+
+        household = household_factory()
+        insulation_quotes = {
+            Element.WALLS: 5_000,
+            Element.GLAZING: 4_000,
+            Element.ROOF: 1_000,
+        }
+        chosen_measures = household.choose_insulation_elements(insulation_quotes, 2)
+
+        assert chosen_measures.keys() == set([Element.ROOF, Element.GLAZING])
+
+        chosen_measures = household.choose_insulation_elements(insulation_quotes, 1)
+
+        assert chosen_measures.keys() == {Element.ROOF}
+
+    def test_installation_of_insulation_measures_improves_element_energy_efficiency_and_epc(
+        self,
+    ) -> None:
+
+        household = household_factory(
+            roof_energy_efficiency=3, walls_energy_efficiency=2, epc=Epc.D
+        )
+        household.install_insulation_elements({Element.ROOF: 1_000})
```

Comment:
How come the argument is a `Dict[Element, Cost]` instead of `Element`?

Reply:
Because I envisage passing the output of `choose_insulation_elements` to multiple downstream functions once households consider insulation as part of a heating system decision -- this one (install them, just uses the keys), but later also one which sums the cost of the insulation measures

Reply:
Ok. If it is just the keys I'd suggest using `.keys()` in the args, but I won't PR withold approval for it.